src/main/java/io/cdap/plugin/couchbase/source/JsonObjectToRecordTransformer.java

```diff
@@ -0,0 +1,240 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.couchbase.source;
+
+import com.couchbase.client.java.document.json.JsonArray;
+import com.couchbase.client.java.document.json.JsonObject;
+import com.google.common.collect.Lists;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.plugin.couchbase.CouchbaseSourceConfig;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+/**
+ * Transforms {@link JsonObject} to {@link StructuredRecord}.
+ */
+public class JsonObjectToRecordTransformer {
+
+  private final CouchbaseSourceConfig config;
+  private final Schema schema;
+
+  public JsonObjectToRecordTransformer(CouchbaseSourceConfig config, Schema schema) {
+    this.config = config;
+    this.schema = schema;
+  }
+
+  /**
+   * Transforms given {@link JsonObject} to {@link StructuredRecord}.
+   *
+   * @param jsonObject JSON object to be transformed.
+   * @return {@link StructuredRecord} that corresponds to the given {@link JsonObject}.
+   */
+  public StructuredRecord transform(JsonObject jsonObject) {
+    JsonObject uniformObject = toUniformJsonObject(jsonObject);
+    return extractRecord(null, uniformObject, schema);
+  }
+
+  /**
+   * Transforms given {@link JsonObject} to it's uniform representation, since returned document set is not required to
+   * be uniform, though it can be. A SELECT statement that specifies a fixed set of attribute (column) names results
+   * in a uniform set of documents and a SELECT statement that specifies the wild card (*) results in a non-uniform
+   * result set. Wild card statements result in an object with single key that matches bucket name.
+   * <p>
+   * See: <a href="https://docs.couchbase.com/server/current/n1ql/n1ql-intro/queriesandresults.html#queries">
+   * N1QL Queries and Results
+   * </a>
+   *
+   * @param jsonObject result document, possibly in a non-uniform representation.
+   * @return uniform representation of the given result document.
+   */
+  private JsonObject toUniformJsonObject(JsonObject jsonObject) {
+    boolean isWildCardQuery = config.getSelectFields().contains("*");
+    if (isWildCardQuery && jsonObject.containsKey(config.getBucket())) {
```

Comment:
the logic here is weird to me, when do each of these conditions happen? For example, when does it contain the bucket, when does it contain just one property, etc?

Can you document the possible inputs and what is output?

Reply:
Examples from the previous comment added to Javadoc.